cpp/src/arrow/util/task_group.cc

```diff
@@ -135,6 +149,18 @@ class ThreadedTaskGroup : public TaskGroup {
       // before cv.notify_one() has returned
       std::unique_lock<std::mutex> lock(mutex_);
       cv_.notify_one();
+      if (completion_future_.has_value()) {
+        // MarkFinished could be slow.  We don't want to call it while we are holding
+        // the lock.
+        // TODO: If optional is thread safe then we can skip this locking entirely
+        auto future = *completion_future_;
+        auto finished = completion_future_->is_finished();
+        auto status = status_;
+        lock.unlock();
+        if (!finished) {
+          future.MarkFinished(status);
```

Comment:
Indeed.  I will make a test case to expose this.

Reply:
I struggled to create a test case that could reproduce this. It seems like it should be reproducible but even adding a slow callback (so MarkFinished took a long time) it was difficult to get a task to be added and then finish just when I wanted it to. I have added some logic however so that this race condition should not be possible any longer.